site/docs/plugins/conversations.md

```diff
@@ -7,1190 +7,1527 @@ next: false
+> Conversations are unique because they introduce a novel concept that you won't find elsewhere in the world.
+> They provide an elegant solution, but you will need to read a bit about how they work before you understand what your code actually does.
 
-When you think about [middleware](../guide/middleware), you will notice that everything is based around a single [context object](../guide/context) per handler.
-This means that you always only handle a single message in isolation.
-It is not easy to write something like "check the text three messages ago" or something.
+Here is a quickstart to let you play around with the plugin before we get to the interesting parts.
 
-**This plugin comes to the rescue:**
-It provides an extremely flexible way to define conversations between your bot and your users.
+:::code-group
 
-Many bot frameworks make you define large configuration objects with steps and stages and jumps and wizard flows and what have you.
-This leads to a lot of boilerplate code, and makes it hard to follow along.
-**This plugin does not work that way.**
-
-Instead, with this plugin, you will use something much more powerful: **code**.
-Basically, you simply define a normal JavaScript function which lets you define how the conversation evolves.
-As the bot and the user talk to each other, the function will be executed statement by statement.
-
-(To be fair, that's not actually how it works under the hood.
-But it is very helpful to think of it that way!
-In reality, your function will be executed a bit differently, but we'll get to that [later](#waiting-for-updates).)
-
-## Simple Example
+```ts [TypeScript]
+import { Bot, type Context } from "grammy";
+import {
+  type Conversation,
+  type ConversationFlavor,
+  conversations,
+  createConversation,
+} from "@grammyjs/conversations";
 
-Before we dive into how you can create conversations, have a look at a short JavaScript example of how a conversation will look.
+const bot = new Bot<ConversationFlavor<Context>>(""); // <-- put your bot token between the "" (https://t.me/BotFather)
+bot.use(conversations());
 
-```js
-async function greeting(conversation, ctx) {
+/** Defines the conversation */
+async function hello(conversation: Conversation, ctx: Context) {
   await ctx.reply("Hi there! What is your name?");
-  const { message } = await conversation.wait();
+  const { message } = await conversation.waitFor("message:text");
   await ctx.reply(`Welcome to the chat, ${message.text}!`);
 }
-```
+bot.use(createConversation(hello));
 
-In this conversation, the bot will first greet the user, and ask for their name.
-Then it will wait until the user sends their name.
-Lastly, the bot welcomes the user to the chat, repeating the name.
+bot.command("enter", async (ctx) => {
+  // Enter the function "hello" you declared.
+  await ctx.conversation.enter("hello");
+});
 
-Easy, right?
-Let's see how it's done!
+bot.start();
+```
```

Comment:
Hmm.. It throws an error on my side:
```ansi
Error in middleware while handling update 141326763 Error: Cannot use conversations without session!
    at https://deno.land/x/grammy_conversations@v1.2.0/conversation.ts:455:19
    at https://deno.land/x/grammy@v1.32.0/composer.ts:141:39
    at https://deno.land/x/grammy@v1.32.0/composer.ts:152:19
    at pass (https://deno.land/x/grammy@v1.32.0/composer.ts:157:12)
    at https://deno.land/x/grammy@v1.32.0/composer.ts:149:15
    at https://deno.land/x/grammy@v1.32.0/composer.ts:149:15
    at https://deno.land/x/grammy@v1.32.0/composer.ts:149:15
    at run (https://deno.land/x/grammy@v1.32.0/composer.ts:171:11)
    at Bot.handleUpdate (https://deno.land/x/grammy@v1.32.0/bot.ts:382:19)
    at Bot.handleUpdates (https://deno.land/x/grammy@v1.32.0/bot.ts:332:28)
No error handler was set!
Set your own error handler with `bot.catch = ...`
Stopping bot
error: Uncaught (in promise) BotError: Error in middleware: Cannot use conversations without session!
```

Here is the code
```ts
import { Bot, type Context } from "https://deno.land/x/grammy@v1.32.0/mod.ts";
import {
  type Conversation,
  type ConversationFlavor,
  conversations,
  createConversation,
} from "https://deno.land/x/grammy_conversations@v1.2.0/mod.ts";

const botToken = Deno.env.get("BOT_TOKEN");
if (!botToken) throw new Error("Missing bot token!");

const bot = new Bot<ConversationFlavor<Context>>(botToken); // <-- put your bot token between the "" (https://t.me/BotFather)
bot.use(conversations());

/** Defines the conversation */
async function hello(conversation: Conversation, ctx: Context) {
  await ctx.reply("Hi there! What is your name?");
  const { message } = await conversation.waitFor("message:text");
  await ctx.reply(`Welcome to the chat, ${message.text}!`);
}
bot.use(createConversation(hello));

bot.command("enter", async (ctx) => {
  // Enter the function "hello" you declared.
  await ctx.conversation.enter("hello");
});

bot.start({
  onStart() {
    console.info("Ready.");
  },
});
```

Reply:
@quadratz This PQ docs is for conversations v2, which you need to import differently. Your code is running conversation v1.2.0, and is not related to the docs in this PQ.

Reply:
Oh, I see. Is there a way for me to test it, or maybe release the rc/beta version?

Reply:
Yes. You can take advantage of deno's ability to import from url, and import from `"https://raw.githubusercontent.com/grammyjs/conversations/refs/heads/main/src/mod.ts"` instead.

Reply:
Great. It's perfectly fine now. Thanks.